torch_frame/nn/encoder/stype_encoder.py

```diff
@@ -685,3 +724,73 @@ def encode_forward(self, feat: TensorData) -> Tensor:
         # -> [batch_size, num_cols, out_channels]
         x = x_lin + self.bias
         return x
+
+
+class TimestampEncoder(StypeEncoder):
+    r"""TimestampEncoder for encoding timestamp stype. The input tensor is
+    expected to have 7 columns representing year, month, day, day of week,
+    hour, minute, and second. The encoding first applies a
+    :class:`torch.nn.Linear` layer to periodic columns and then multiplies
+    different sine and cosine frequencies to them. The result embedding is
+    then concatenated to the only linear column, year. A final
+    :class:`torch.nn.Linear` is applied the the concatenated output.
+    """
+    supported_stypes = {stype.timestamp}
+
+    def __init__(
+        self,
+        out_channels: Optional[int] = None,
+        stats_list: Optional[List[Dict[StatType, Any]]] = None,
+        stype: Optional[stype] = None,
+        post_module: Optional[Module] = None,
+        na_strategy: Optional[NAStrategy] = None,
+    ):
+        super().__init__(out_channels, stats_list, stype, post_module,
+                         na_strategy)
+
+    def init_modules(self):
+        super().init_modules()
+        self.periodic = Linear(stype.timestamp.timestamp_num_cols - 1,
+                               stype.timestamp.timestamp_num_cols - 1)
+        min_year = torch.tensor([
+            self.stats_list[i][StatType.YEAR_RANGE][0]
+            for i in range(len(self.stats_list))
+        ])
+        self.register_buffer("min_year", min_year)
+        max_year = torch.tensor([
+            self.stats_list[i][StatType.YEAR_RANGE][1]
+            for i in range(len(self.stats_list))
+        ])
+        self.register_buffer("max_year", max_year)
+        self.mult_constants = [torch.pi, torch.pi * 2, torch.pi * 3]
+        self.linear = Linear(
+            2 * len(self.mult_constants) *
+            (stype.timestamp.timestamp_num_cols - 1) + 1, self.out_channels)
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        super().reset_parameters()
+        self.linear.reset_parameters()
+        self.periodic.reset_parameters()
+
+    def encode_forward(self, feat: Tensor) -> Tensor:
+        device = feat.device
+        feat[:, :,
+             TIME_TO_INDEX['YEAR']] = (feat[:, :, 0] - self.min_year.to(device)
+                                       ) / (self.max_year.to(device) -
+                                            self.min_year.to(device) + 1)
```

Comment:
Better formatting. Avoid min_year device transfer

Reply:
For year, can we not normalize?
just subtract the current year by the min year and apply Transformer's positional encoding (see [here](https://machinelearningmastery.com/a-gentle-introduction-to-positional-encoding-in-transformer-models-part-1/)). You may find some official implementation.